softwareconfiguration/model/src/main/java/ai/libs/softwareconfiguration/model/ComponentInstance.java

```diff
@@ -1,329 +1,327 @@
+		other.annotations.entrySet().forEach(x -> this.annotations.put(x.getKey(), x.getValue()));
+	}
+
+	/**
+	 * Constructor for creating a <code>ComponentInstance</code> for a particular <code>Component</code>.
+	 *
+	 * @param component
+	 *            The component that is grounded.
+	 * @param parameterValues
+	 *            A map containing the parameter values of this grounding.
+	 * @param satisfactionOfRequiredInterfaces
+	 *            The refinement of the required interfaces.
+	 */
+	public ComponentInstance(@JsonProperty("component") final Component component, @JsonProperty("parameterValues") final Map<String, String> parameterValues,
+			@JsonProperty("satisfactionOfRequiredInterfaces") final Map<String, ComponentInstance> satisfactionOfRequiredInterfaces) {
+		super();
+		this.component = component;
+		this.parameterValues = parameterValues;
+		this.satisfactionOfRequiredInterfaces = satisfactionOfRequiredInterfaces;
+	}
+
+	/**
+	 * @return The <code>Component</code> to this <code>ComponentInstance</code>.
+	 */
+	public Component getComponent() {
+		return this.component;
+	}
+
+	/**
+	 * @return The parameters and how their values were set.
+	 */
+	public Map<String, String> getParameterValues() {
+		return this.parameterValues;
+	}
+
+	/**
+	 * @return The set of parameters of which the values have been set explicitly.
+	 */
+	public Collection<Parameter> getParametersThatHaveBeenSetExplicitly() {
+		if (this.parameterValues == null) {
+			return new ArrayList<>();
+		}
+		return this.getComponent().getParameters().stream().filter(p -> this.parameterValues.containsKey(p.getName())).collect(Collectors.toList());
+	}
+
+	/**
+	 * @return The set of parameters of which the values have not been set explicitly.
+	 */
+	public Collection<Parameter> getParametersThatHaveNotBeenSetExplicitly() {
+		return SetUtil.difference(this.component.getParameters(), this.getParametersThatHaveBeenSetExplicitly());
+	}
+
+	/**
+	 * @param param
+	 *            The parameter for which the value shall be returned.
+	 * @return The value of the parameter.
+	 */
+	public String getParameterValue(final Parameter param) {
+		return this.getParameterValue(param.getName());
+	}
+
+	/**
+	 * @param paramName
+	 *            The name of the parameter for which the value is requested.
+	 * @return The value of the parameter with the given name.
+	 */
+	public String getParameterValue(final String paramName) {
+		return this.parameterValues.get(paramName);
+	}
+
+	/**
+	 * @return This method returns a mapping of interface IDs to component instances.
+	 */
+	public Map<String, ComponentInstance> getSatisfactionOfRequiredInterfaces() {
+		return this.satisfactionOfRequiredInterfaces;
+	}
+
+	/**
+	 * @return A collection of all components contained (recursively) in this <code>ComponentInstance</code>.
+	 */
+	public Collection<Component> getContainedComponents() {
```

Comment:
Do we really want to have this method in a ComponentInstance or does this rather belong to a util class?